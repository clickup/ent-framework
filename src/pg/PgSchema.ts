import type { Query } from "../abstract/Query";
import { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  InsertInput,
  LoadByInput,
  Row,
  SelectByInput,
  SelectInput,
  Table,
  UniqueKey,
  UpdateInput,
} from "../types";
import { PgQueryCount } from "./PgQueryCount";
import { PgQueryDelete } from "./PgQueryDelete";
import { PgQueryExists } from "./PgQueryExists";
import { PgQueryIDGen } from "./PgQueryIDGen";
import { PgQueryInsert } from "./PgQueryInsert";
import { PgQueryLoad } from "./PgQueryLoad";
import { PgQueryLoadBy } from "./PgQueryLoadBy";
import { PgQuerySelect } from "./PgQuerySelect";
import { PgQuerySelectBy } from "./PgQuerySelectBy";
import { PgQueryUpdate } from "./PgQueryUpdate";
import { PgQueryUpsert } from "./PgQueryUpsert";

export class PgSchema<
  TTable extends Table,
  TUniqueKey extends UniqueKey<TTable>,
> extends Schema<TTable, TUniqueKey> {
  idGen(): Query<string> {
    return new PgQueryIDGen(this);
  }

  insert(input: InsertInput<TTable>): Query<string | null> {
    return new PgQueryInsert(this, input);
  }

  upsert(input: InsertInput<TTable>): Query<string> {
    return new PgQueryUpsert(this, input);
  }

  update(id: string, input: UpdateInput<TTable>): Query<boolean> {
    return new PgQueryUpdate(this, id, input);
  }

  delete(id: string): Query<boolean> {
    return new PgQueryDelete(this, id);
  }

  load(id: string): Query<Row<TTable> | null> {
    return new PgQueryLoad(this, id);
  }

  loadBy(input: LoadByInput<TTable, TUniqueKey>): Query<Row<TTable> | null> {
    return new PgQueryLoadBy(this, input);
  }

  selectBy(
    input: SelectByInput<TTable, TUniqueKey>,
  ): Query<Array<Row<TTable>>> {
    return new PgQuerySelectBy(this, input);
  }

  select(input: SelectInput<TTable>): Query<Array<Row<TTable>>> {
    return new PgQuerySelect(this, input);
  }

  count(input: CountInput<TTable>): Query<number> {
    return new PgQueryCount(this, input);
  }

  exists(input: ExistsInput<TTable>): Query<boolean> {
    return new PgQueryExists(this, input);
  }
}
