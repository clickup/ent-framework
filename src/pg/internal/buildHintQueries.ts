import type { Hints } from "../../types";

export const RAW_PREPEND_HINT = "";

/**
 * Builds query prologue queries for the given hints.
 *
 * In the resulting compound queries, the returned `queries` will become a part
 * of the debug query text, and queriesDefault will be prepended and omitted
 * from the debug query text.
 *
 * Also, if there is a special hint with key = "", the its value is appended as
 * it is to the very beginning of the compound query sent. You can e.g. pass
 * pg_hint_plan extension hints there.
 */
export function buildHintQueries(
  hintsDefault: Readonly<Hints> = {},
  hints: Readonly<Hints> = {},
): [rawPrepend: string, queriesDefault: string[], queries: string[]] {
  const queriesDefault: string[] = [];
  const queries: string[] = [];

  const rawPrepend = hints[RAW_PREPEND_HINT] ?? "";

  for (const k in hintsDefault) {
    const v = hintsDefault[k];
    if (k === RAW_PREPEND_HINT) {
      continue;
    } else if (v === null || v === undefined) {
      // Engine default or non-set.
      continue;
    } else if (hints[k] !== undefined) {
      // User sets this hint to some different value (or resets it).
      continue;
    } else {
      queriesDefault.push(buildHintQuery(k, v));
    }
  }

  for (const k in hints) {
    const v = hints[k];
    if (k === RAW_PREPEND_HINT) {
      continue;
    } else if (v === null || v === undefined) {
      // Engine default or non-set.
      continue;
    } else {
      queries.push(buildHintQuery(k, v));
    }
  }

  return [rawPrepend, queriesDefault, queries];
}

function buildHintQuery(k: string, v: string): string {
  return k.toLowerCase() === "transaction"
    ? `SET LOCAL ${k} ${v}`
    : `SET LOCAL ${k} TO ${v}`;
}
