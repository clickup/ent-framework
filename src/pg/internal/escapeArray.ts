import { escapeString } from "./escapeString";

/**
 * Escapes an array of strings.
 */
export function escapeArray(
  obj: Array<string | null> | null | undefined,
): string {
  return obj === null || obj === undefined
    ? "NULL"
    : escapeString(
        "{" +
          obj
            .map((v) =>
              v === null
                ? "NULL"
                : `"${v.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"`,
            )
            .join(",") +
          "}",
      );
}
