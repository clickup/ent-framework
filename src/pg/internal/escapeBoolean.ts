/**
 * Escapes a boolean as PG string literal.
 */
export function escapeBoolean(v: boolean | null | undefined): string {
  return v === null || v === undefined ? "NULL" : v ? "true" : "false";
}
