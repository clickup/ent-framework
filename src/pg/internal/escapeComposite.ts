import { escapeString } from "./escapeString";
import { parseCompositeRow } from "./parseCompositeRow";

/**
 * PostgreSQL doesn't allow comparison like `WHERE (a, b) = '(1,2)'` - it throws
 * "Input of anonymous composite types is not implemented" error. So to compare,
 * we have to convert the stringified row representation to ROW() notation
 * manually: `WHERE (a, b) = ROW('1', '2')`
 *
 * Notice that we don't work with ROWs consisting of 1 element; instead, we
 * treat them as the element itself. I.e. instead of emitting "(123)" or
 * "ROW(123)", we always emit just "123".
 *
 * - "1" => "1"
 * - "(1)" => "1"
 * - "(1,2)" => "ROW('1','2')"
 */
export function escapeComposite(v: string | null | undefined): string {
  if (v === null || v === undefined) {
    return "NULL";
  }

  const parts =
    v.startsWith("(") && v.endsWith(")") ? parseCompositeRow(v) : [v];
  const list = parts.map((v) => escapeString(v)).join(",");
  return parts.length > 1 ? `ROW(${list})` : list;
}
