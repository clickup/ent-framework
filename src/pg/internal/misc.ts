import type {
  ClientErrorKind,
  ClientErrorPostAction,
} from "../../abstract/ClientError";

/**
 * Some errors affect the logic of choosing another Client when a query retry is
 * requested (controlled via ClientError). Mostly, those are the situation when
 * a PG node goes down.
 *
 * Notice that, if a master goes down, we always need to find another master
 * (and retry hard hoping is a switchover, and it will appear on another node).
 */
export const CLIENT_ERROR_PREDICATES: Array<
  (error: { code: string; message: string }) =>
    | false
    | {
        abbreviation: string;
        postAction: {
          ifMaster: ClientErrorPostAction;
          ifReplica: ClientErrorPostAction;
          reportConnectionIssue: boolean;
        };
        kind: ClientErrorKind;
        comment: string;
      }
> = [
  ({ code }) =>
    code === "42P01" && {
      abbreviation: "undefined_table",
      postAction: {
        ifMaster: "rediscover-cluster",
        ifReplica: "rediscover-cluster",
        reportConnectionIssue: false,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "For single-queries: table doesn't exist or disappeared (e.g. the Shard was relocated to another Island).",
    },
  ({ code }) =>
    code === "42704" && {
      abbreviation: "undefined_object",
      postAction: {
        ifMaster: "rediscover-cluster",
        ifReplica: "rediscover-cluster",
        reportConnectionIssue: false,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "For batched queries: table doesn't exist or disappeared (e.g. the Shard was relocated to another Island).",
    },
  ({ code }) =>
    code === "25006" && {
      abbreviation: "read_only_sql_transaction",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "rediscover-island",
        reportConnectionIssue: false,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "A write happened in a read-only Client (probably the Client's role was changed from master to replica due to a failover/switchover).",
    },
  ({ code }) =>
    code === "57P01" && {
      abbreviation: "admin_shutdown",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment: "The database is shutting down by an administrator.",
    },
  ({ code }) =>
    code === "57P02" && {
      abbreviation: "crash_shutdown",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment: "The database is crashed and is shutting down.",
    },
  ({ message }) =>
    message === "Connection terminated unexpectedly" && {
      abbreviation: "connection_terminated",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "unknown-server-state", // !!!
      comment:
        "Node-postgres connection terminated unexpectedly (from client.js).",
    },
  ({ code }) =>
    code === "ECONNREFUSED" && {
      abbreviation: "ECONNREFUSED",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment: "Node TCP library connect error.",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message.includes("server conn crashed") && {
      abbreviation: "server_conn_crashed",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "unknown-server-state",
      comment: "PG is terminated cruelly (e.g. by SIGKILL or SIGSEGV).",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message === "query_wait_timeout" && {
      abbreviation: "query_wait_timeout",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "PG went down a few seconds ago, but a PgBouncer connection is still open, so the query sent to it times out waiting. After some time, PgBouncer will stop emitting this error and start emitting server_login_retry on new connections instead.",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message.includes("server_login_retry") && {
      abbreviation: "server_login_retry",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "PG has been down for quite some, so PgBouncer emits this error on each new connection immediately. Or, PG server has just been shut down, and PgBouncer's server pool is empty (fast-fail use case).",
    },
  ({ code, message }) =>
    code === "08P01" && // protocol_violation
    message === "pgbouncer cannot connect to server" && {
      abbreviation: "cannot_connect_to_server",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        "Older PgBouncer versions emit this error instead of server_login_retry.",
    },
  ({ code }) =>
    code === "08P01" && {
      // protocol_violation
      abbreviation: "protocol_violation_catch_all",
      postAction: {
        ifMaster: "rediscover-island",
        ifReplica: "choose-another-client",
        reportConnectionIssue: true,
      },
      kind: "data-on-server-is-unchanged",
      comment:
        'Sometimes PgBouncer emit unrelated error codes when PG server crashes (e.g. "invalid server parameter").',
    },
];
