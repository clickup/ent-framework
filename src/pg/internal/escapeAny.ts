import { escapeArray } from "./escapeArray";
import { escapeBoolean } from "./escapeBoolean";
import { escapeDate } from "./escapeDate";
import { escapeString } from "./escapeString";

/**
 * Tries its best to escape the value according to its type.
 *
 * Try to not use this function; although it protects against SQL injections,
 * it's not aware of the actual field type, so it e.g. cannot prevent a bigint
 * overflow SQL error.
 */
export function escapeAny(v: unknown): string {
  return v === null || v === undefined
    ? "NULL"
    : typeof v === "number"
      ? v.toString()
      : typeof v === "boolean"
        ? escapeBoolean(v)
        : v instanceof Date
          ? escapeDate(v)
          : v instanceof Array
            ? escapeArray(v)
            : escapeString(v as string | null | undefined);
}
