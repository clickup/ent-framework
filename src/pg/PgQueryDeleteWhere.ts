import type { QueryAnnotation } from "../abstract/QueryAnnotation";
import { QueryBase } from "../abstract/QueryBase";
import type { Schema } from "../abstract/Schema";
import { stringHash } from "../internal/misc";
import type { DeleteWhereInput, Table } from "../types";
import { ID } from "../types";
import type { PgClient } from "./PgClient";
import { PgRunner } from "./PgRunner";

export class PgQueryDeleteWhere<TTable extends Table> extends QueryBase<
  TTable,
  DeleteWhereInput<TTable>,
  string[],
  PgClient
> {
  /** @ignore */
  readonly RUNNER_CLASS = PgRunnerDeleteWhere;
}

class PgRunnerDeleteWhere<TTable extends Table> extends PgRunner<
  TTable,
  DeleteWhereInput<TTable>,
  string[]
> {
  static override readonly IS_WRITE = true;
  private builder;

  readonly op = "DELETE_WHERE";
  readonly maxBatchSize = 100;
  readonly default = [];

  // This runner doesn't support batching.
  runBatch = undefined;

  constructor(schema: Schema<TTable>, client: PgClient) {
    super(schema, client);
    this.builder = this.createWhereBuilder({
      prefix: this.fmt("DELETE FROM %T "),
      suffix: this.fmt(` RETURNING %PK AS ${ID}`),
    });
  }

  override key(input: DeleteWhereInput<TTable>): string {
    // Coalesce equal delete queries.
    const json = JSON.stringify(input);
    return stringHash(json);
  }

  async runSingle(
    input: DeleteWhereInput<TTable>,
    annotations: QueryAnnotation[],
  ): Promise<string[]> {
    if (!(input[ID] instanceof Array)) {
      throw Error(
        `Field ${ID} must be an array of IDs in ${this.op} query (for safety)`,
      );
    }

    const sql =
      this.builder.prefix + this.builder.func(input) + this.builder.suffix;
    const rows = await this.clientQuery<{ [ID]: string }>(
      sql,
      annotations,
      input[ID].length,
    );
    return rows.map((row) => row[ID]);
  }
}
