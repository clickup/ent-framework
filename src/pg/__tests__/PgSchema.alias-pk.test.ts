import type { Shard } from "../../abstract/Shard";
import { MASTER } from "../../abstract/Shard";
import { nullthrows } from "../../internal/misc";
import { ID } from "../../types";
import { PgQueryDeleteWhere } from "../PgQueryDeleteWhere";
import { PgSchema } from "../PgSchema";
import type { TestPgClient } from "./test-utils";
import { recreateTestTables, shardRun, testCluster } from "./test-utils";

const schema = new PgSchema(
  'pg-schema.alias-pk"table',
  {
    user_id: { type: ID, autoInsert: "id_gen()" },
    name: { type: String },
  },
  ["user_id"],
);

let shard: Shard<TestPgClient>;
let master: TestPgClient;

beforeEach(async () => {
  await recreateTestTables([
    {
      CREATE: [
        `CREATE TABLE %T(
          user_id bigint NOT NULL PRIMARY KEY,
          name text NOT NULL
        )`,
      ],
      SCHEMA: schema,
      SHARD_AFFINITY: [],
    },
  ]);

  shard = await testCluster.randomShard();
  master = await shard.client(MASTER);
});

test("single ops", async () => {
  const id1 = await shardRun(
    shard,
    schema.insert({ user_id: "1", name: "n1" }),
  );
  master.toMatchSnapshot();
  expect(id1).toEqual("1");

  const id2 = await shardRun(
    shard,
    schema.upsert({ user_id: "2", name: "n2" }),
  );
  master.toMatchSnapshot();

  const id3 = await shardRun(shard, schema.insert({ name: "n3" }));
  master.toMatchSnapshot();

  {
    const res = await shardRun(shard, schema.update(nullthrows(id1), {})); // no DB query is sent here
    master.toMatchSnapshot();
    expect(res).toEqual(true);
  }

  {
    const res = await shardRun(
      shard,
      schema.upsert({ user_id: "101", name: "n11" }),
    );
    master.toMatchSnapshot();
    expect(res).toEqual("101");
  }

  {
    const res = await shardRun(shard, schema.delete(nullthrows(id1)));
    master.toMatchSnapshot();
    expect(res).toEqual(true);
  }

  {
    const res = await shardRun(
      shard,
      new PgQueryDeleteWhere(schema, { id: [id2!], $literal: ["1=1"] }),
    );
    master.toMatchSnapshot();
    expect(res).toEqual([id2]);
  }

  {
    const res = await shardRun(shard, schema.load(id3!));
    master.toMatchSnapshot();
    expect(res).toMatchObject({ id: id3, name: "n3" });
  }

  {
    const res = await shardRun(shard, schema.loadBy({ user_id: id3! }));
    master.toMatchSnapshot();
    expect(res).toMatchObject({ id: id3, name: "n3" });
  }

  {
    const res = await shardRun(shard, schema.count({ user_id: id3! }));
    master.toMatchSnapshot();
    expect(res).toEqual(1);
  }

  {
    const res = await shardRun(shard, schema.exists({ user_id: id3! }));
    master.toMatchSnapshot();
    expect(res).toStrictEqual(true);
  }

  {
    const res = await shardRun(
      shard,
      schema.select({ where: { user_id: id3! }, limit: 10 }),
    );
    master.toMatchSnapshot();
    expect(res).toMatchObject([{ user_id: id3 }]);
  }
});
