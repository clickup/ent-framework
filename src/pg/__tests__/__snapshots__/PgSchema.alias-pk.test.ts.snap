// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single ops 1`] = `
"
INSERT INTO "pg-schema.alias-pk""table" (name, user_id) VALUES
  ('n1', '1') ON CONFLICT DO NOTHING RETURNING user_id AS id
----
INSERT INTO "pg-schema.alias-pk""table" (name, user_id) VALUES
  ('?', '?') ON CONFLICT DO NOTHING RETURNING user_id AS id
"
`;

exports[`single ops 2`] = `
"
INSERT INTO "pg-schema.alias-pk""table" (user_id, name) VALUES
  ('2', 'n2')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
----
INSERT INTO "pg-schema.alias-pk""table" (user_id, name) VALUES
  ('?', '?')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
"
`;

exports[`single ops 3`] = `
"
INSERT INTO "pg-schema.alias-pk""table" (name, user_id) VALUES
  ('n3', id_gen()) ON CONFLICT DO NOTHING RETURNING user_id AS id
----
INSERT INTO "pg-schema.alias-pk""table" (name, user_id) VALUES
  ('?', id_gen()) ON CONFLICT DO NOTHING RETURNING user_id AS id
"
`;

exports[`single ops 4`] = `""`;

exports[`single ops 5`] = `
"
INSERT INTO "pg-schema.alias-pk""table" (user_id, name) VALUES
  ('101', 'n11')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
----
INSERT INTO "pg-schema.alias-pk""table" (user_id, name) VALUES
  ('?', '?')
  ON CONFLICT (user_id) DO UPDATE SET name=EXCLUDED.name RETURNING user_id AS id
"
`;

exports[`single ops 6`] = `
"
DELETE FROM "pg-schema.alias-pk""table" WHERE user_id IN('1') RETURNING user_id AS id
----
DELETE FROM "pg-schema.alias-pk""table" WHERE user_id IN('?') RETURNING user_id AS id
"
`;

exports[`single ops 7`] = `
"
DELETE FROM "pg-schema.alias-pk""table" WHERE user_id IN('2') AND (1=1) RETURNING user_id AS id
----
DELETE FROM "pg-schema.alias-pk""table" WHERE user_id IN('?') AND (?=?) RETURNING user_id AS id
"
`;

exports[`single ops 8`] = `
"
SELECT user_id, name, user_id AS id FROM "pg-schema.alias-pk""table"
  WHERE user_id IN('<id>')
----
SELECT user_id, name, user_id AS id FROM "pg-schema.alias-pk""table"
  WHERE user_id IN('?')
"
`;

exports[`single ops 9`] = `
"
SELECT user_id, name, user_id AS id FROM "pg-schema.alias-pk""table"
  WHERE user_id='<id>'
----
SELECT user_id, name, user_id AS id FROM "pg-schema.alias-pk""table"
  WHERE user_id='?'
"
`;

exports[`single ops 10`] = `
"
SELECT COUNT(1) AS count FROM "pg-schema.alias-pk""table"
  WHERE user_id='<id>'
----
SELECT COUNT(?) AS count FROM "pg-schema.alias-pk""table"
  WHERE user_id='?'
"
`;

exports[`single ops 11`] = `
"
SELECT EXISTS (SELECT true FROM "pg-schema.alias-pk""table"
  WHERE user_id='<id>')
----
SELECT EXISTS (SELECT true FROM "pg-schema.alias-pk""table"
  WHERE user_id='?')
"
`;

exports[`single ops 12`] = `
"
SELECT user_id, name, user_id AS id FROM "pg-schema.alias-pk""table"
  WHERE user_id='<id>'
  LIMIT 10
----
SELECT user_id, name, user_id AS id FROM "pg-schema.alias-pk""table"
  WHERE user_id='?'
  LIMIT ?
"
`;
